nemo/collections/asr/modules/sortformer_modules.py

```diff
@@ -12,68 +12,145 @@
 from nemo.core.classes.module import NeuralModule
+from nemo.utils import logging
 
 __all__ = ['SortformerModules']
 
 
+def concat_and_pad(embs: List[torch.Tensor], lengths: List[torch.Tensor]):
+    """Concatenates lengths[i] first embeddings of embs[i], and pads the rest elements with zeros.
+    Args:
+        embs: List of embeddings Tensors of (batch_size, n_frames of i-th, emb_dim) shape
+        lengths: List of lengths Tensors of (batch_size,) shape
+
+    Returns:
+        output: concatenated embeddings Tensor of (batch_size, , emb_dim) shape
+        total_lengths: output lengths Tensor of (batch_size,) shape
+    """
+
+    if len(embs) != len(lengths):
+        raise ValueError(
+            f"Mismatch in embedding length. Length of embs {len(embs)} and " f"lengths {len(lengths)} must be equal."
+        )
+    device = embs[0].device
+    dtype = embs[0].dtype
+    batch_size, emb_dim = embs[0].shape[0], embs[0].shape[2]
+
+    total_lengths = torch.sum(torch.stack(lengths), dim=0)
+    max_length = total_lengths.max().item()
+
+    output = torch.zeros(batch_size, max_length, emb_dim, device=device, dtype=dtype)
+    start_indices = torch.zeros(batch_size, dtype=torch.int64, device=device)
+
+    for emb, length in zip(embs, lengths):
+        end_indices = start_indices + length
+        for batch_idx in range(batch_size):
+            output[batch_idx, start_indices[batch_idx] : end_indices[batch_idx]] = emb[batch_idx, : length[batch_idx]]
+        start_indices = end_indices
+
+    return output, total_lengths
+
+
+@dataclass
+class StreamingSortformerState:
+    spkcache = None  # speaker cache to store embeddings from start
+    spkcache_lengths = None
+    spkcache_preds = None  # speaker cache predictions
+    fifo = None  # to save the embedding from the latest chunks
+    fifo_lengths = None
+    fifo_preds = None
+    spk_perm = None
+
+
 class SortformerModules(NeuralModule, Exportable):
     """
     A class including auxiliary functions for Sortformer models.
     This class contains and will contain the following functions that performs streaming features,
-    and any neural layers that are not included in the NeMo neural modules (e.g. Transformer, Fast-Conformer).
+    and any neural layers that are not included in the NeMo neural modules
+    (e.g. Transformer, Fast-Conformer).
     """
 
     def init_weights(self, m):
         """Init weights for linear layers."""
-        if type(m) == nn.Linear:
+        if isinstance(m, nn.Linear):
             torch.nn.init.xavier_uniform_(m.weight)
             m.bias.data.fill_(0.01)
 
     def __init__(
         self,
         num_spks: int = 4,
-        hidden_size: int = 192,
         dropout_rate: float = 0.5,
         fc_d_model: int = 512,
         tf_d_model: int = 192,
+        subsampling_factor: int = 8,
+        spkcache_len: int = 188,
+        fifo_len: int = 0,
+        chunk_len: int = 376,
+        spkcache_refresh_rate: int = 1,
```

Comment:
Would these variables (e.g. spkcache_refresh_rate, spkcache_sil_frames_per_spk, causal_attn_rate, scores_boost_latest, min_pos_scores_rate etc) be explained in the config? A description for them would be helpful for users to correctly set these values